File: data/Data_week_3/d1/generate_data.py
import pandas as pd
from faker import Faker
from datetime import datetime, timedelta
import random
import os

# Initialize Faker
fake = Faker()

# Set random seed for reproducibility
random.seed(42)
fake.seed_instance(42)

# Helper function to generate valid phone number
def generate_phone():
    return '0' + fake.numerify('#########')

# Helper function to generate invalid phone for junk data (5% chance)
def generate_phone_with_junk():
    if random.random() < 0.05:  # 5% junk data
        junk_types = random.choice(['short', 'letters', 'long'])
        if junk_types == 'short':
            return fake.numerify('#####')
        elif junk_types == 'letters':
            return fake.lexify('??????????')
        else:
            return '0' + fake.numerify('###########')
    else:
        return generate_phone()

# 1. Generate branches (5 branches)
branches_data = []
branch_names = ['Chi Nhánh Trung Tâm', 'Chi Nhánh Sài Gòn', 'Chi Nhánh Hà Nội', 'Chi Nhánh Đà Nẵng', 'Chi Nhánh Cần Thơ']
for i in range(5):
    branches_data.append({
        'branch_id': i+1,
        'branch_name': branch_names[i],
        'address': fake.address(),
        'phone_number': generate_phone_with_junk()
    })
branches_df = pd.DataFrame(branches_data)

# 2. Generate categories (10 categories)
categories_data = []
category_names = ['Thuốc giảm đau', 'Kháng sinh', 'Thuốc chống trầm cảm', 'Thuốc chống dị ứng', 'Thuốc chống loạn thần',
                  'Thuốc tim mạch', 'Thuốc da liễu', 'Thuốc nội tiết', 'Thuốc tiêu hóa', 'Thuốc hô hấp']
for i in range(10):
    categories_data.append({
        'category_id': i+1,
        'category_name': category_names[i],
        'description': fake.text(100)
    })
categories_df = pd.DataFrame(categories_data)

# 3. Generate medicines (500 medicines)
medicines_data = []
units = ['tablet', 'vial', 'bottle', 'box']
for i in range(500):
    category_id = random.randint(1, 10)
    name = fake.company() + ' ' + fake.word()  # Simple medicine name
    if random.random() < 0.05:  # Junk data
        name = '' if random.random() < 0.5 else name
    brand = fake.company()
    unit = random.choice(units)
    if random.random() < 0.05:  # Junk data
        unit = random.choice(['kg', 'ml', 'pack'])  # Invalid units
    description = fake.text(50)
    created_at = fake.date_time_between(start_date='-1y', end_date='now')
    medicines_data.append({
        'medicine_id': i+1,
        'category_id': category_id,
        'name': name,
        'brand': brand,
        'unit': unit,
        'description': description,
        'created_at': created_at
    })
medicines_df = pd.DataFrame(medicines_data)

# 4. Generate batches (2000 batches)
batches_data = []
batch_count_per_med = {}
for med_id in range(1, 501):
    batch_count_per_med[med_id] = random.randint(3, 5)

batch_med_ids = []
for med_id, count in batch_count_per_med.items():
    batch_med_ids.extend([med_id] * count)

random.shuffle(batch_med_ids)
batch_med_ids = batch_med_ids[:2000]  # Ensure exactly 2000

for i in range(2000):
    medicine_id = batch_med_ids[i]
    batch_number = fake.unique.numerify('BATCH-#####')
    manufacturing_date = fake.date_between(start_date='-2y', end_date='today')
    expiry_date = manufacturing_date + timedelta(days=random.randint(365, 1460))  # 1-4 years later
    import_price = round(random.uniform(10, 1000), 2)
    batches_data.append({
        'batch_id': i+1,
        'medicine_id': medicine_id,
        'batch_number': batch_number,
        'manufacturing_date': manufacturing_date,
        'expiry_date': expiry_date,
        'import_price': import_price
    })
batches_df = pd.DataFrame(batches_data)

# 5. Generate inventory (2000 batches * 5 branches = 10000 rows)
inventory_data = []
for batch_id in range(1, 2001):
    for branch_id in range(1, 6):
        quantity = random.randint(0, 1000)
        if random.random() < 0.05:  # Junk data
            quantity = -random.randint(1, 100)  # Negative quantity
        last_updated = fake.date_time_between(start_date='-1y', end_date='now')
        inventory_data.append({
            'inventory_id': len(inventory_data)+1,
            'branch_id': branch_id,
            'batch_id': batch_id,
            'quantity': quantity,
            'last_updated': last_updated
        })
inventory_df = pd.DataFrame(inventory_data)

# 6. Generate pharmacists (20 pharmacists, 4 per branch)
pharmacists_data = []
for branch_id in range(1, 6):
    for _ in range(4):
        full_name = fake.name()
        license_number = fake.unique.numerify('LIC-#####')
        username = fake.unique.user_name()
        password = fake.password(length=8)
        role = random.choice(['STAFF', 'ADMIN'])
        phone_number = generate_phone_with_junk()
        pharmacists_data.append({
            'pharmacist_id': len(pharmacists_data)+1,
            'branch_id': branch_id,
            'full_name': full_name,
            'license_number': license_number,
            'username': username,
            'password': password,
            'role': role
            # Note: phone_number not in schema, but perhaps add if needed, but schema doesn't have
        })
pharmacists_df = pd.DataFrame(pharmacists_data)

# 7. Generate customers (1000 customers)
customers_data = []
for i in range(1000):
    full_name = fake.name()
    phone_number = generate_phone_with_junk()
    loyalty_points = random.randint(0, 10000)
    customers_data.append({
        'customer_id': i+1,
        'full_name': full_name,
        'phone_number': phone_number,
        'loyalty_points': loyalty_points
    })
customers_df = pd.DataFrame(customers_data)

# 8. Generate invoices (7000 invoices)
invoices_data = []
for i in range(7000):
    branch_id = random.randint(1, 5)
    pharmacist_id = random.randint(1, 20)
    customer_id = random.randint(1, 1000) if random.random() > 0.2 else None  # 20% no customer
    total_amount = round(random.uniform(50, 5000), 2)
    sale_date = fake.date_time_between(start_date='-1y', end_date='now')
    invoices_data.append({
        'invoice_id': i+1,
        'branch_id': branch_id,
        'pharmacist_id': pharmacist_id,
        'customer_id': customer_id,
        'total_amount': total_amount,
        'sale_date': sale_date
    })
invoices_df = pd.DataFrame(invoices_data)

# 9. Generate invoice_details (for each invoice, 1-5 details)
invoice_details_data = []
for inv in invoices_df.itertuples():
    num_details = random.randint(1, 5)
    branch_id = inv.branch_id
    # Get available batches in this branch with quantity > 0
    available_batches = inventory_df[(inventory_df['branch_id'] == branch_id) & (inventory_df['quantity'] > 0)]['batch_id'].unique().tolist()
    if len(available_batches) < num_details:
        # If not enough, take all available
        selected_batches = available_batches
    else:
        selected_batches = random.sample(available_batches, num_details)
    for batch_id in selected_batches:
        quantity = random.randint(1, 10)
        # Get import_price from batches
        import_price = batches_df.loc[batches_df['batch_id'] == batch_id, 'import_price'].values[0]
        unit_price = round(import_price * 1.5, 2)  # Markup
        invoice_details_data.append({
            'detail_id': len(invoice_details_data)+1,
            'invoice_id': inv.invoice_id,
            'batch_id': batch_id,
            'quantity': quantity,
            'unit_price': unit_price
        })
invoice_details_df = pd.DataFrame(invoice_details_data)

# Export to CSV
dfs = {
    'branches': branches_df,
    'categories': categories_df,
    'medicines': medicines_df,
    'batches': batches_df,
    'inventory': inventory_df,
    'pharmacists': pharmacists_df,
    'customers': customers_df,
    'invoices': invoices_df,
    'invoice_details': invoice_details_df
}

for name, df in dfs.items():
    df.to_csv(f'{name}.csv', index=False, encoding='utf-8-sig')  # utf-8-sig for Vietnamese characters

print("Data generation complete. CSV files created.")

File: data/Data_week_3/d1/generate_data.js
const { faker } = require('@faker-js/faker');
const createCsvWriter = require('csv-writer').createObjectCsvWriter;

// Constants
const NUM_BRANCHES = 5;
const NUM_CATEGORIES = 10;
const NUM_MEDICINES = 500;
const NUM_BATCHES = 2000; // 4 per medicine
const NUM_PHARMACISTS = 50; // 10 per branch
const NUM_CUSTOMERS = 1000;
const NUM_INVOICES = 7000;

// Units
const UNITS = ['tablet', 'vial', 'bottle', 'box'];

// Generate branches
function generateBranches() {
  const branches = [];
  for (let i = 0; i < NUM_BRANCHES; i++) {
    branches.push({
      branch_id: i + 1,
      branch_name: faker.company.name() + ' Pharmacy',
      address: faker.location.streetAddress(),
      phone_number: '0' + faker.string.numeric(9)
    });
  }
  return branches;
}

// Generate categories
function generateCategories() {
  const categories = [];
  for (let i = 0; i < NUM_CATEGORIES; i++) {
    categories.push({
      category_id: i + 1,
      category_name: faker.commerce.department(),
      description: faker.lorem.sentence()
    });
  }
  return categories;
}

// Generate medicines
function generateMedicines() {
  const medicines = [];
  for (let i = 0; i < NUM_MEDICINES; i++) {
    const isBad = Math.random() < 0.05; // 5% bad data
    const unit = isBad && Math.random() < 0.5 ? faker.lorem.word() : faker.helpers.arrayElement(UNITS);
    const name = isBad && Math.random() < 0.5 ? '' : faker.commerce.productName();
    medicines.push({
      medicine_id: i + 1,
      category_id: faker.number.int({ min: 1, max: NUM_CATEGORIES }),
      name: name,
      brand: faker.company.name(),
      unit: unit,
      description: faker.lorem.sentence(),
      created_at: faker.date.past().toISOString().slice(0, 19).replace('T', ' ')
    });
  }
  return medicines;
}

// Generate batches
function generateBatches() {
  const batches = [];
  let batchId = 1;
  for (let medId = 1; medId <= NUM_MEDICINES; medId++) {
    for (let b = 0; b < 4; b++) { // 4 batches per medicine
      const expiryYear = faker.number.int({ min: 2024, max: 2027 });
      const manufacturingDate = faker.date.between({ from: new Date('2020-01-01'), to: new Date('2023-12-31') });
      const expiryDate = faker.date.between({ from: new Date(`${expiryYear}-01-01`), to: new Date(`${expiryYear}-12-31`) });
      batches.push({
        batch_id: batchId++,
        medicine_id: medId,
        batch_number: faker.string.alphanumeric(10).toUpperCase(),
        manufacturing_date: manufacturingDate.toISOString().split('T')[0],
        expiry_date: expiryDate.toISOString().split('T')[0],
        import_price: faker.number.float({ min: 1, max: 100, precision: 0.01 })
      });
    }
  }
  return batches;
}

// Generate inventory
function generateInventory(batches) {
  const inventory = [];
  let invId = 1;
  batches.forEach(batch => {
    for (let branchId = 1; branchId <= NUM_BRANCHES; branchId++) {
      const isBad = Math.random() < 0.05;
      const quantity = isBad && Math.random() < 0.5 ? faker.number.int({ min: -100, max: -1 }) : faker.number.int({ min: 0, max: 1000 });
      inventory.push({
        inventory_id: invId++,
        branch_id: branchId,
        batch_id: batch.batch_id,
        quantity: quantity,
        last_updated: faker.date.recent({ days: 1 }).toISOString().slice(0, 19).replace('T', ' ')
      });
    }
  });
  return inventory;
}

// Generate pharmacists
function generatePharmacists() {
  const pharmacists = [];
  for (let i = 0; i < NUM_PHARMACISTS; i++) {
    const branchId = ((i % NUM_BRANCHES) + 1);
    pharmacists.push({
      pharmacist_id: i + 1,
      branch_id: branchId,
      full_name: faker.person.fullName(),
      license_number: faker.string.alphanumeric(10),
      username: faker.internet.username(),
      password: faker.internet.password(),
      role: faker.helpers.arrayElement(['STAFF', 'ADMIN'])
    });
  }
  return pharmacists;
}

// Generate customers
function generateCustomers() {
  const customers = [];
  for (let i = 0; i < NUM_CUSTOMERS; i++) {
    const isBad = Math.random() < 0.05;
    const phone = isBad && Math.random() < 0.5 ? faker.phone.number() : '0' + faker.string.numeric(9);
    customers.push({
      customer_id: i + 1,
      full_name: faker.person.fullName(),
      phone_number: phone,
      loyalty_points: faker.number.int({ min: 0, max: 1000 })
    });
  }
  return customers;
}

// Generate invoices and details
function generateInvoicesAndDetails(batches, pharmacists, customers) {
  const invoices = [];
  const invoiceDetails = [];
  let invId = 1;
  let detId = 1;
  for (let i = 0; i < NUM_INVOICES; i++) {
    const branchId = faker.number.int({ min: 1, max: NUM_BRANCHES });
    const pharmacist = pharmacists.find(p => p.branch_id === branchId);
    const customerId = faker.number.int({ min: 1, max: NUM_CUSTOMERS });
    const saleDate = faker.date.recent({ days: 365 }).toISOString().slice(0, 19).replace('T', ' ');
    invoices.push({
      invoice_id: invId,
      branch_id: branchId,
      pharmacist_id: pharmacist.pharmacist_id,
      customer_id: customerId,
      total_amount: 0, // will calculate
      sale_date: saleDate
    });

    let total = 0;
    const numDetails = faker.number.int({ min: 1, max: 5 });
    for (let d = 0; d < numDetails; d++) {
      const batch = faker.helpers.arrayElement(batches);
      const quantity = faker.number.int({ min: 1, max: 10 });
      const unitPrice = batch.import_price * faker.number.float({ min: 1.1, max: 2.0, precision: 0.01 }); // markup
      total += quantity * unitPrice;
      invoiceDetails.push({
        detail_id: detId++,
        invoice_id: invId,
        batch_id: batch.batch_id,
        quantity: quantity,
        unit_price: unitPrice
      });
    }
    invoices[i].total_amount = total.toFixed(2);
    invId++;
  }
  return { invoices, invoiceDetails };
}

// Write to CSV
function writeToCSV(filename, records, headers) {
  const csvWriter = createCsvWriter({
    path: filename,
    header: headers
  });
  csvWriter.writeRecords(records).then(() => console.log(`${filename} written`));
}

// Main
async function main() {
  const branches = generateBranches();
  const categories = generateCategories();
  const medicines = generateMedicines();
  const batches = generateBatches();
  const inventory = generateInventory(batches);
  const pharmacists = generatePharmacists();
  const customers = generateCustomers();
  const { invoices, invoiceDetails } = generateInvoicesAndDetails(batches, pharmacists, customers);

  // Write CSVs
  writeToCSV('branches.csv', branches, [
    { id: 'branch_id', title: 'branch_id' },
    { id: 'branch_name', title: 'branch_name' },
    { id: 'address', title: 'address' },
    { id: 'phone_number', title: 'phone_number' }
  ]);

  writeToCSV('categories.csv', categories, [
    { id: 'category_id', title: 'category_id' },
    { id: 'category_name', title: 'category_name' },
    { id: 'description', title: 'description' }
  ]);

  writeToCSV('medicines.csv', medicines, [
    { id: 'medicine_id', title: 'medicine_id' },
    { id: 'category_id', title: 'category_id' },
    { id: 'name', title: 'name' },
    { id: 'brand', title: 'brand' },
    { id: 'unit', title: 'unit' },
    { id: 'description', title: 'description' },
    { id: 'created_at', title: 'created_at' }
  ]);

  writeToCSV('batches.csv', batches, [
    { id: 'batch_id', title: 'batch_id' },
    { id: 'medicine_id', title: 'medicine_id' },
    { id: 'batch_number', title: 'batch_number' },
    { id: 'manufacturing_date', title: 'manufacturing_date' },
    { id: 'expiry_date', title: 'expiry_date' },
    { id: 'import_price', title: 'import_price' }
  ]);

  writeToCSV('inventory.csv', inventory, [
    { id: 'inventory_id', title: 'inventory_id' },
    { id: 'branch_id', title: 'branch_id' },
    { id: 'batch_id', title: 'batch_id' },
    { id: 'quantity', title: 'quantity' },
    { id: 'last_updated', title: 'last_updated' }
  ]);

  writeToCSV('pharmacists.csv', pharmacists, [
    { id: 'pharmacist_id', title: 'pharmacist_id' },
    { id: 'branch_id', title: 'branch_id' },
    { id: 'full_name', title: 'full_name' },
    { id: 'license_number', title: 'license_number' },
    { id: 'username', title: 'username' },
    { id: 'password', title: 'password' },
    { id: 'role', title: 'role' }
  ]);

  writeToCSV('customers.csv', customers, [
    { id: 'customer_id', title: 'customer_id' },
    { id: 'full_name', title: 'full_name' },
    { id: 'phone_number', title: 'phone_number' },
    { id: 'loyalty_points', title: 'loyalty_points' }
  ]);

  writeToCSV('invoices.csv', invoices, [
    { id: 'invoice_id', title: 'invoice_id' },
    { id: 'branch_id', title: 'branch_id' },
    { id: 'pharmacist_id', title: 'pharmacist_id' },
    { id: 'customer_id', title: 'customer_id' },
    { id: 'total_amount', title: 'total_amount' },
    { id: 'sale_date', title: 'sale_date' }
  ]);

  writeToCSV('invoice_details.csv', invoiceDetails, [
    { id: 'detail_id', title: 'detail_id' },
    { id: 'invoice_id', title: 'invoice_id' },
    { id: 'batch_id', title: 'batch_id' },
    { id: 'quantity', title: 'quantity' },
    { id: 'unit_price', title: 'unit_price' }
  ]);

  console.log('All CSVs generated');
}

main();